istioctl/cmd/proxystatus.go

```diff
@@ -150,8 +150,12 @@ func newKubeClientWithRevision(kubeconfig, configContext string, revision string
 	return kube.NewCLIClient(kube.NewClientConfigForRestConfig(rc), revision)
 }
 
-func newKubeClient(kubeconfig, configContext string) (kube.CLIClient, error) {
-	return newKubeClientWithRevision(kubeconfig, configContext, "")
+func newKubeClient(kubeconfig, configContext string) error {
```

Comment:
Oof just realized there's a potential bug here; if the kubeconfig or context changes, the value of the returned kubeClient won't change. I don't know the istioctl code well enough to know if multiple kubeConfigs are 
used/supported, but this could be a pretty subtle bug if it is. I wonder if it's even worth it to memoize the client at this point? 🤔 because doubt this right would require additional complexity and I'm not sure the benefit is all that useful as I see more of the code...in fact, if most of the kubeClients initialized here are only in scope for the duration of their function, they'll be garbage collected as soon as the function is over. With a package level var, it lives on for the duration of the programs execution (which is probably pretty short given it's a CLI). I know the sync.Once was initially my suggestion, but seeing the fully fleshed out implementation, I'm not sure i'm confident the complexity is worth it. What are your thoughts @MorrisLaw?

Reply:
Another thought: if we only set the kubeconfig and context once per cobra cmd invocation, we should plumb it through via a struct of some sort so that we actually can memoize it without making the getKubeClient function call so unstable.

Reply:
> Oof just realized there's a potential bug here; if the kubeconfig or context changes, the value of the returned kubeClient won't change.

I tried to read and understand the code about this point here before making the PR and came to the conclusion that it should only be set once (as you stated later on in your next comment). You can see how it gets set [here ](https://github.com/istio/istio/blob/09917748493f1e29389d6c5eb9dee0f99db796fc/istioctl/cmd/root.go#L152-L156) actually. So if I understand how cobra works, then this rootCmd should be the only thing setting those values. Unless I'm missing something of course.

Reply:
> Another thought: if we only set the kubeconfig and context once per cobra cmd invocation, we should plumb it through via a struct of some sort so that we actually can memoize it without making the getKubeClient function call so unstable.

I'm open to this. I'm not sure how the code will look yet, but let me get back to you after I play around with it a bit.

One question I'd have though is what is the benefit? Or, rather, what makes `getKubeClient` unstable? I'm not quite clear on that part.

Thank you for reviewing this @keithmattix !

Reply:
That's a good point. Trying to follow the code and I don't see any reason why we even _have_ to do that. It's possible this was a pattern that was just carried over.. :thinking:

That could probably be removed tbh. Wdyt?

I traced it back to when it was first added: https://github.com/istio/istio/blob/ee321d48c8f2b9242793ed0b3309b5d044aa3e93/istioctl/cmd/proxystatus.go#L128

It looks like it was once called `newSDSExecClient` when it was added initially. Then it was changed to `newEnvoyClient` then changed to `newKubeClient` with a few more modifications until now. But `kubeconfig` was carried through the whole way.

It also appears that `kubeconfig` was refactored out of main.go: https://github.com/istio/istio/commit/87e427881a6809bea5d9abfd4b9aa03cca23e93b

It's possible the passing in of that `kubeconfig` variable was carried over but not actually needed for how it's used now? 

I can't definitively say for sure of course as I'm new to this codebase myself. This is all basically conjecture at this point. You may have more of an opinion on this than I do. I'm open to other suggestions.

Reply:
Either way, as it currently stands, I don't see `kubeconfig` getting reset anywhere within the `cmd` package. It appears that it's just the use of the global variable we see getting initially read in via rootCmd.

Reply:
Then yeah I'd say remove the param  and see if tests still pass. If so, the sync.Once should still be viable. Ideally, the context is only set once as well and that param can be removed too

Reply:
Yeah that makes sense. Just updated it and tests pass locally 👍